### PR TITLE
(almost) Always use `ObligationCtxt` when dealing with canonical queries

### PR DESCRIPTION
**Pull request by compiler-errors**
Hope this is a step in the right direction. cc rust-lang/types-team#50.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_traits/src/type_op.rs [hunk 1]
```diff
@@ -61,89 +57,74 @@ pub fn type_op_ascribe_user_type_with_span<'a, 'tcx: 'a>(
...
-            .at(&ObligationCause::dummy_with_span(self.span), self.param_env)
-            .relate(a, variance, b)?
-            .into_value_registering_obligations(self.infcx, self.fulfill_cx);
-        Ok(())
+        Ok(self.ocx.register_infer_ok_obligations(
+            self.ocx
+                .infcx
+                .at(&ObligationCause::dummy_with_span(self.span), self.param_env)
+                .eq(a, b)?,
+        ))
```
Comment (lcnr) [inline compiler/rustc_traits/src/type_op.rs:95]: use `self.ocx.equate` directly here

File: compiler/rustc_traits/src/type_op.rs [hunk 2]
```diff
@@ -199,28 +179,26 @@ fn type_op_eq<'tcx>(
...
+    tcx.infer_ctxt().enter_canonical_trait_query(&canonicalized, |ocx, key| {
         let (param_env, Eq { a, b }) = key.into_parts();
-        infcx
-            .at(&ObligationCause::dummy(), param_env)
-            .eq(a, b)?
-            .into_value_registering_obligations(infcx, fulfill_cx);
+        ocx.register_infer_ok_obligations(
+            ocx.infcx.at(&ObligationCause::dummy(), param_env).eq(a, b)?,
+        );
         Ok(())
```
Comment (lcnr) [inline compiler/rustc_traits/src/type_op.rs:187]: `ocx.equate`

File: compiler/rustc_traits/src/type_op.rs [hunk 3]
```diff
@@ -199,28 +179,26 @@ fn type_op_eq<'tcx>(
...
 ) -> Fallible<T>
 where
     T: fmt::Debug + TypeFoldable<'tcx> + Lift<'tcx>,
 {
     let (param_env, Normalize { value }) = key.into_parts();
     let Normalized { value, obligations } =
-        infcx.at(&ObligationCause::dummy(), param_env).normalize(value)?;
-    fulfill_cx.register_predicate_obligations(infcx, obligations);
+        ocx.infcx.at(&ObligationCause::dummy(), param_env).normalize(value)?;
+    ocx.register_obligations(obligations);
```
Comment (lcnr) [inline compiler/rustc_traits/src/type_op.rs:201]: `ocx.normalize`

File: compiler/rustc_traits/src/type_op.rs [hunk 4]
```diff
@@ -256,12 +234,11 @@ fn type_op_subtype<'tcx>(
...
-    tcx.infer_ctxt().enter_canonical_trait_query(&canonicalized, |infcx, fulfill_cx, key| {
+    tcx.infer_ctxt().enter_canonical_trait_query(&canonicalized, |ocx, key| {
         let (param_env, Subtype { sub, sup }) = key.into_parts();
-        infcx
-            .at(&ObligationCause::dummy(), param_env)
-            .sup(sup, sub)?
-            .into_value_registering_obligations(infcx, fulfill_cx);
+        ocx.register_infer_ok_obligations(
+            ocx.infcx.at(&ObligationCause::dummy(), param_env).sup(sup, sub)?,
+        );
```
Comment (lcnr) [inline compiler/rustc_traits/src/type_op.rs:241]: `ocx.sup` (doesn't exist, but should)

**Review by lcnr — COMMENTED** [review]
:heart: 

further improvements, then r=me

File: compiler/rustc_traits/src/type_op.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_traits/src/type_op.rs:201]: ocx.normalize has different semantics than infcx.normalize though -- it doesn't return an error when it failed to normalize.

**Comment by nnethercote** [comment]
This might just be noise, but I'm not certain. Either way, it's just one of the secondary benchmarks, so a regression wouldn't matter much.

@rustbot label: +perf-regression-triaged

**Pull request merged**